core/templates/domain/topic/editable-topic-backend-api.service.ts

```diff
@@ -16,233 +16,298 @@
+  'skill_question_count_dict': Object;
+  'skill_id_to_rubrics_dict': Object;
+  'classroom_url_fragment': string;
+}
+
+interface FetchStoriesBackendResponse {
+  'canonical_story_summary_dicts': Object;
+}
+
+interface FetchSubtopicPageBackendResponse {
+  'subtopic_page': Object;
+}
+
+interface DeleteTopicBackendResponse {
+  status: Object;
+}
+
+interface UpdateTopicBackendResponse {
+  'topic_dict': Object;
+  'skill_id_to_description_dict': Object;
+  'skill_id_to_rubrics_dict': Object
+}
+
+interface DoesTopicWithUrlFragmentExistBackendResponse {
+  'topic_url_fragment_exists': Object
+}
 
-      $http.get(storiesDataUrl).then(function(response) {
-        var canonicalStorySummaries = angular.copy(
-          response.data.canonical_story_summary_dicts);
-        if (successCallback) {
-          successCallback(canonicalStorySummaries);
-        }
-      }, function(errorResponse) {
-        if (errorCallback) {
-          errorCallback(errorResponse.data);
-        }
+interface DoesTopicWithNameExistBackendResponse {
+  'topic_name_exists': Object
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class EditableTopicBackendApiService {
+  constructor(
+    private http: HttpClient,
+    private urlInterpolationService: UrlInterpolationService) {}
+
+  private _fetchTopic(
+      topicId: string,
+      successCallback: (value?: Object | PromiseLike<Object>) => void,
+      errorCallback: (reason?: string) => void): void {
+    let topicDataUrl = this.urlInterpolationService.interpolateUrl(
+      AppConstants.EDITABLE_TOPIC_DATA_URL_TEMPLATE, {
+        topic_id: topicId
       });
-    };
 
-    var _fetchSubtopicPage = function(
-        topicId, subtopicId, successCallback, errorCallback) {
-      var subtopicPageDataUrl = UrlInterpolationService.interpolateUrl(
-        SUBTOPIC_PAGE_EDITOR_DATA_URL_TEMPLATE, {
-          topic_id: topicId,
-          subtopic_id: subtopicId.toString()
+    this.http.get<FetchTopicBackendResponse>(
+      topicDataUrl).toPromise().then((response) => {
+      if (successCallback) {
+        // The response is passed as a dict with 2 fields and not as 2
+        // parameters, because the successCallback is called as the resolve
+        // callback function in $q in fetchTopic(), and according to its
+        // documentation (https://docs.angularjs.org/api/ng/service/$q),
+        // resolve or reject can have only a single parameter.
+        successCallback({
+          topicDict: response.topic_dict,
+          groupedSkillSummaries: response.grouped_skill_summary_dicts,
+          skillIdToDescriptionDict: response.skill_id_to_description_dict,
+          skillQuestionCountDict: {
+            ...response.skill_question_count_dict
+          },
+          skillIdToRubricsDict: response.skill_id_to_rubrics_dict,
+          classroomUrlFragment: response.classroom_url_fragment,
         });
+      }
+    }, (errorResponse) => {
+      if (errorCallback) {
```

Comment:
We can remove these 'if' statements because we are always passing in success and error callbacks.

Reply:
Done